src/server/game/Movement/MovementPacketSender.cpp

```diff
@@ -0,0 +1,367 @@
+    { SMSG_SPLINE_SET_RUN_BACK_SPEED,    SMSG_FORCE_RUN_BACK_SPEED_CHANGE,       MSG_MOVE_SET_RUN_BACK_SPEED },
+    { SMSG_SPLINE_SET_SWIM_SPEED,        SMSG_FORCE_SWIM_SPEED_CHANGE,           MSG_MOVE_SET_SWIM_SPEED },
+    { SMSG_SPLINE_SET_SWIM_BACK_SPEED,   SMSG_FORCE_SWIM_BACK_SPEED_CHANGE,      MSG_MOVE_SET_SWIM_BACK_SPEED },
+    { SMSG_SPLINE_SET_TURN_RATE,         SMSG_FORCE_TURN_RATE_CHANGE,            MSG_MOVE_SET_TURN_RATE },
+    { SMSG_SPLINE_SET_FLIGHT_SPEED,      SMSG_FORCE_FLIGHT_SPEED_CHANGE,         MSG_MOVE_SET_FLIGHT_SPEED },
+    { SMSG_SPLINE_SET_FLIGHT_BACK_SPEED, SMSG_FORCE_FLIGHT_BACK_SPEED_CHANGE,    MSG_MOVE_SET_FLIGHT_BACK_SPEED },
+    { SMSG_SPLINE_SET_PITCH_RATE,        SMSG_FORCE_PITCH_RATE_CHANGE,           MSG_MOVE_SET_PITCH_RATE },
+};
+
+void MovementPacketSender::SendSpeedChangeToMover(Unit* unit, UnitMoveType mtype, float newRate)
+{
+    Player* mover = unit->GetPlayerMovingMe();
+    if (!mover)
+    {
+        TC_LOG_ERROR("entities.unit", "MovementPacketSender::SendSpeedChangeToMover: Incorrect use of the function. It was called on a unit controlled by the server!");
+        return;
+    }
+
+    float newSpeedFlat = newRate * (mover->IsControlledByPlayer() ? playerBaseMoveSpeed[mtype] : baseMoveSpeed[mtype]); // this line is a fucking mess. what if the unit is a creature MCed by a player? this whole speed rate thing needs to die. In the meantime: use mover or unit ?
+    uint32 mCounter = unit->GetMovementCounterAndInc();
+    PlayerMovementPendingChange pendingChange;
+    pendingChange.movementCounter = mCounter;
+    pendingChange.newValue = newSpeedFlat;
+    pendingChange.movementChangeType = MovementPacketSender::GetChangeTypeByMoveType(mtype);
+    unit->PushPendingMovementChange(pendingChange);
+
+    WorldPacket data;
+    data.Initialize(moveTypeToOpcode[mtype][1], mtype != MOVE_RUN ? 8 + 4 + 4 : 8 + 4 + 1 + 4);
+    data << unit->GetPackGUID();
+    data << mCounter;
+    if (mtype == MOVE_RUN)
+        data << uint8(1);                               // unknown byte added in 2.1.0
+    data << newSpeedFlat;
+    mover->GetSession()->SendPacket(&data);
+}
+
+MovementChangeType MovementPacketSender::GetChangeTypeByMoveType(UnitMoveType moveType)
+{
+    switch (moveType)
+    {
+        case MOVE_WALK:         return SPEED_CHANGE_WALK;
+        case MOVE_RUN:          return SPEED_CHANGE_RUN;
+        case MOVE_RUN_BACK:     return SPEED_CHANGE_RUN_BACK;
+        case MOVE_SWIM:         return SPEED_CHANGE_SWIM;
+        case MOVE_SWIM_BACK:    return SPEED_CHANGE_SWIM_BACK;
+        case MOVE_TURN_RATE:    return RATE_CHANGE_TURN;
+        case MOVE_FLIGHT:       return SPEED_CHANGE_FLIGHT_SPEED;
+        case MOVE_FLIGHT_BACK:  return SPEED_CHANGE_FLIGHT_BACK_SPEED;
+        case MOVE_PITCH_RATE:   return RATE_CHANGE_PITCH;
+        default:
+            throw new std::runtime_error("MovementPacketSender::SendSpeedChangeToMover Unsupported UnitMoveType");
```

Comment:
ASSERT(condition, "nice message that you can even %s format arguments", "pass");